linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala

```diff
@@ -2002,6 +2022,20 @@ private[emitter] class FunctionEmitter(sjsGen: SJSGen) {
       })
     }
 
+    private def withTempJSVar(value: Tree)(makeBody: Transient => js.Tree)(
+        implicit env: Env, pos: Position): js.Tree = {
+      withTempJSVar(transformExpr(value, value.tpe), value.tpe)(makeBody)
+    }
+
+    private def withTempJSVar(value: js.Tree, tpe: Type)(
+        makeBody: Transient => js.Tree)(
+        implicit pos: Position): js.Tree = {
+      val varIdent = newSyntheticVar()
+      val varDef = genLet(varIdent, mutable = false, value)
+      val body = makeBody(Transient(JSVarRef(varIdent, mutable = false)(tpe)))
+      js.Block(varDef, body)
+    }
+
```

Comment:
I have, and that's what I initially tried, but it is very tricky for these. They introduce another thing where the CoreJSLib would instantiate the class and therefore create a static module dependency from `j.l.Object` to it. Except this time it's not the *only* place where they get instantiated. Combining that with the conditional reachability became hard.

Reply:
Yeah, I see. Managing the CoreJSLib's requirements seems tricky when we need to track module dependencies.

What I'm a bit bothered by is that it seems that we probably also generate more code like this (then if we'd put it into helpers).

I have one more a bit crazy idea: What if we put the helpers in the linker private library as static methods (on a `WrapHelpers` class) and simply call them? We'd need to teach the info analyzer this, but that's IMO actually even slightly better than reaching exactly what is inside the method.

Reply:
I don't know about your crazy idea. It would work, of course. But I would prefer to keep the linker-private-library as small as possible, and contain only *necessary* stuff. Otherwise, it will be too easy to indulge in putting more and more things in it. For example, why not implement integer division/modulo in there?

(Also in user-space we cannot avoid an addition `AsInstanceOf(e, Throwable)` in the `then` branch, unfortunately.)

Reply:
> Hmmm... Fair enough. I wonder if we could teach the optimizer to optimize these.

This is tough. We'll need flow-based information for that. We should do it one day (because it comes up all the time in pattern matching) but this is really out of scope for this PR.

> > Otherwise, it will be too easy to indulge in putting more and more things in it. For example, why not implement integer division/modulo in there?
> 
> I think that would be a good thing. IMO it's better if more code is written in Scala than in hand-written trees.
> 
> In fact, I suspect that if we allow the optimizer to substitute these implementations, a lot of specific optimizations might become superseded by the general purpose optimizations.

Only if we actually always inline these methods. Currently we don't do that for integer division, for example, but we can elide the complex function when the divisor is known not to be 0.

I don't think this is the way to go, in general. We have the user-space library (including javalib and scalalib, but not linker-private-lib), which is built on top an IR. And we have an IR, which provides a contract to the user-space libraries, and is built on top of JavaScript (trees). Implementing some of the IR on top of pseudo-user-space (the linker-private-lib) that itself relies on the IR is already dodgy. We manage by having the pseudo-user-space carefully crafted so that the features it implements do not rely on the IR nodes for those features. But the more we put there, the harder this becomes.

Reply:
Ahahaha, fair enough. Yeah. I didn't think about the circularity of the whole thing. Indeed that's not a good idea.